Improve module type resolver system so that it would be usable for Alpaka modules

makortel opened the pull request:
#### PR description:

This PR contains changes to the generic module type resolver system (https://github.com/cms-sw/cmssw/pull/37731, https://github.com/cms-sw/cmssw/pull/39391) in the framework so that it would be usable for a specific implementation for Alpaka modules (that comes in a subsequent PR). I explored a few approaches on the implementation, and settled in one that makes most of its work in the python configuration (as opposed to C++ code).

The changes include in the python configuration side
* Allow `UnavailableAccelerator` `edm::Exception` to be thrown in python (extending #40315) 
* Add `ModuleTypeResolver` functionality to `ProcessAccelerator`:
  * `ProcessAccelerator` can return (via `moduleTypeResolver()` member function) a "type resolver" object (based on the accelerators available in the machine) that is used to 
      * specify a `ModuleTypeResolver` plugin for C++ code with an accompanied PSet
      * customize each module PSet; must limit to a PSet specifying the variant ("backend") of the module with untracked parameters

and in the C++ side
* Use abstract factory pattern for module type resolver in order to use information from the module PSet in the type resolver, in a way that the PSet is parsed only once per module instance
* Module makers are now cached by their resolved module type in order to allow the type resolver return different type name per module instance
* `ModuleTypeResolverMaker` is constructed in `EventProcessor`, and percolated to relevant ED and ES module factories

#### PR validation:

Framework unit tests pass (except for `TestIOPoolInputNoParentDictionary` that needs an improvement done in a separate PR).

makortel committed:
Add plugin factory for ModuleTypeResolverMaker

makortel committed:
Percolate the ModuleTypeResolverMaker from EventProcessor to ModuleRegistry

makortel committed:
Add ModuleTypeResolver functionality to ProcessAccelerator

makortel committed:
Cache the module makers by resolved module type

This change allows the type resolver to return different type name per
module instance.

makortel committed:
Use abstract factory pattern for module type resolver

In order to be able to use module PSet information in the type
resolver, and parse the PSet only once per module instance.

makortel committed:
Make ConfigurableTestTypeResolverMaker a plugin and use it in a cmsRun test

makortel committed:
Allow UnavailableAccelerator edm::Exception to be thrown in python

cmsbuild commented:
A new Pull Request was created by  @makortel (Matti Kortelainen) for master.

It involves the following packages:

- FWCore/Framework (**core**)
- FWCore/Integration (**core**)
- FWCore/ParameterSet (**core**)
- FWCore/PythonParameterSet (**core**)
- FWCore/TestProcessor (**core**)


@cmsbuild, @smuzaffar, @Dr15Jones, @makortel can you please review it and eventually sign? Thanks.
@missirol, @wddgit this is something you requested to watch as well.
@perrotta, @dpiparo, @rappoccio you are the release manager for this.

cms-bot commands are listed <a href="http://cms-sw.github.io/cms-bot-cmssw-cmds.html">here</a>

makortel commented:
@cmsbuild, please test with #40384

makortel commented:
@Dr15Jones Would you have time to review before holidays?

makortel commented on FWCore/Framework/interface/ModuleTypeResolverBase.h:
```diff
@@ -35,9 +38,24 @@ namespace edm {
+     * This function creates an implementation of the ModuleTypeResolverBase class based on the module PSet.
+     *
+     * The return value reflects that the implementation
+     * ModuleTypeResolverBase may cache the Resolver objects if it
```
This should probably be
```suggestion
     * ModuleTypeResolverMaker may cache the Resolver objects if it
```

makortel replied:
Done

makortel commented on FWCore/Framework/interface/ModuleTypeResolverMakerFactory.h:
```diff
@@ -0,0 +1,16 @@
+  class ModuleTypeResolverMaker;
+  class ParameterSet;
+
+  using ModuleTypeResolverMakerFactory = edmplugin::PluginFactory<ModuleTypeResolverMaker*(edm::ParameterSet const&)>;
```
The `edm::ParameterSet` parameter is strictly speaking a remnant of earlier stages of the development. It is used in some of the tests, but the model with which I'd proceed with Alpaka would not use it (instead the necessary information on the variant, i.e. backend, is delivered as part of the module PSet).

I left the parameter so that the mechanism would be more general, but now I'm not sure if it would really be useful, so I could also remove it.

makortel replied:
Removed the `edm::ParameterSet` parameter.

makortel commented on FWCore/ParameterSet/python/Config.py:
```diff
@@ -2136,11 +2184,41 @@ def __init__(self, **kargs):
+        def pluginAndConfiguration(self):
+            return ("TestModuleTypeResolver", self._pset)
+
+        def setModuleBackend(self, module):
```
I think this member function should be renamed. `setModuleVariant()` comes to my mind. Any better ideas?

makortel replied:
Now named as `setModuleVariant()` (better suggestions are still welcome)

makortel commented on FWCore/Framework/test/stubs/TestTypeResolverPlugins.cc:
```diff
@@ -0,0 +1,32 @@
+          throw edm::Exception(edm::errors::Configuration) << "Variant must be 'other' or 'cpu'. Got " << variant;
+        }
+      }
+      return std::make_shared<ConfigurableTestTypeResolver>(variant);
```
I wonder now if I should try to make use of the caching of the `TypeResolver` object here.

makortel replied:
Added caching.

makortel reviewed (commented):
Some remarks of my own.

cmsbuild commented:
-1

**Failed Tests**:   Build 
**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-cb2a0a/29732/summary.html
**COMMIT**: c2aa487a3a01a77b95f0738c8d9edd8830668131
**CMSSW**: CMSSW_13_0_X_2022-12-20-2300/el8_amd64_gcc11
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week0/cms-sw/cmssw/40383/29732/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Build

I found compilation error when building: 

<pre>>> Compiling  /data/cmsbld/jenkins/workspace/ib-run-pr-tests/CMSSW_13_0_X_2022-12-20-2300/src/Mixing/Base/src/BMixingModule.cc
>> Compiling  /data/cmsbld/jenkins/workspace/ib-run-pr-tests/CMSSW_13_0_X_2022-12-20-2300/src/Mixing/Base/src/PileUp.cc
>> Compiling  /data/cmsbld/jenkins/workspace/ib-run-pr-tests/CMSSW_13_0_X_2022-12-20-2300/src/Mixing/Base/src/PileupRandomNumberGenerator.cc
>> Compiling  /data/cmsbld/jenkins/workspace/ib-run-pr-tests/CMSSW_13_0_X_2022-12-20-2300/src/Mixing/Base/src/SecondaryEventProvider.cc
/data/cmsbld/jenkins/workspace/ib-run-pr-tests/CMSSW_13_0_X_2022-12-20-2300/src/Mixing/Base/src/SecondaryEventProvider.cc: In constructor 'edm::SecondaryEventProvider::SecondaryEventProvider(std::vector<edm::ParameterSet>&, edm::ProductRegistry&, std::shared_ptr<edm::ProcessConfiguration>)':
/data/cmsbld/jenkins/workspace/ib-run-pr-tests/CMSSW_13_0_X_2022-12-20-2300/src/Mixing/Base/src/SecondaryEventProvider.cc:55:9: error: no matching function for call to 'edm::WorkerManager::WorkerManager(std::shared_ptr<edm::ActivityRegistry>, edm::ExceptionToActionTable&)'
   55 |         workerManager_(std::make_shared<ActivityRegistry>(), *exceptionToActionTable_) {
      |         ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from /data/cmsbld/jenkins/workspace/ib-run-pr-tests/CMSSW_13_0_X_2022-12-20-2300/src/Mixing/Base/src/SecondaryEventProvider.h:4,
                 from /data/cmsbld/jenkins/workspace/ib-run-pr-tests/CMSSW_13_0_X_2022-12-20-2300/src/Mixing/Base/src/SecondaryEventProvider.cc:2:
/data/cmsbld/jenkins/workspace/ib-run-pr-tests/CMSSW_13_0_X_2022-12-20-2300/src/FWCore/Framework/interface/WorkerManager.h:46:5: note: candidate: 'edm::WorkerManager::WorkerManager(std::shared_ptr<edm::ModuleRegistry>, std::shared_ptr<edm::ActivityRegistry>, const edm::ExceptionToActionTable&)'
</pre>

makortel commented:
@cmsbuild, please test

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-40383/33515

- This PR adds an extra 12KB to repository

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  FWCore/ParameterSet/python/Config.py  modified in PR(s): #37996

cmsbuild commented:
Pull request #40383 was updated. @makortel, @smuzaffar, @civanch, @Dr15Jones, @mdhildreth can you please check and sign again.

cmsbuild commented:
+1

**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-cb2a0a/29746/summary.html
**COMMIT**: 6fcd8b7c6278b190db0c68a9c2937bc3628b9fc3
**CMSSW**: CMSSW_13_0_X_2022-12-20-2300/el8_amd64_gcc11
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week0/cms-sw/cmssw/40383/29746/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Comparison Summary



**Summary**:
 - No significant changes to the logs found
 - Reco comparison results: 4 differences found in the comparisons
 - DQMHistoTests: Total files compared: 49
 - DQMHistoTests: Total histograms compared: 3555748
 - DQMHistoTests: Total failures: 154
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3555572
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 48 files compared)
 - Checked 211 log files, 162 edm output root files, 49 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_13_0_X_2022-12-20-2300+cb2a0a/54781/triggerResults): no differences found

cmsbuild commented:
+1

**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-cb2a0a/29746/summary.html
**COMMIT**: 6fcd8b7c6278b190db0c68a9c2937bc3628b9fc3
**CMSSW**: CMSSW_13_0_X_2022-12-20-2300/el8_amd64_gcc11
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week0/cms-sw/cmssw/40383/29746/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Comparison Summary



**Summary**:
 - No significant changes to the logs found
 - Reco comparison results: 4 differences found in the comparisons
 - DQMHistoTests: Total files compared: 49
 - DQMHistoTests: Total histograms compared: 3555748
 - DQMHistoTests: Total failures: 154
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3555572
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 48 files compared)
 - Checked 211 log files, 162 edm output root files, 49 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_13_0_X_2022-12-20-2300+cb2a0a/54781/triggerResults): no differences found

makortel committed:
Add test for configuration hash staying unchanged

makortel committed:
Enable module type resolver for TestProcessor

makortel commented:
Addressed my own review  comments, added a test to show that the approach continues to preserve configuration hash, and extended the type resolver functionality to `TestProcessor` (for which I had earlier a comment in code).

cmsbuild commented:
-code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-40383/33527

- This PR adds an extra 256KB to repository

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  FWCore/ParameterSet/python/Config.py  modified in PR(s): #37996

Code check has found code style and quality issues which could be resolved by applying following patch(s)

- **code-format**:
https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-40383/33527/code-format.patch
e.g. `curl -k https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-40383/33527/code-format.patch | patch -p1`
You can also run `scram build code-format` to apply code format directly

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-40383/33528

- This PR adds an extra 116KB to repository

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  FWCore/ParameterSet/python/Config.py  modified in PR(s): #37996

cmsbuild commented:
Pull request #40383 was updated. @smuzaffar, @civanch, @Dr15Jones, @makortel, @mdhildreth, @cmsbuild can you please check and sign again.

makortel commented:
@cmsbuild, please test

cmsbuild commented:
+1

**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-cb2a0a/29763/summary.html
**COMMIT**: 93678a2f018ca14153ea9d9a204795c0e713f19e
**CMSSW**: CMSSW_13_0_X_2022-12-22-1100/el8_amd64_gcc11
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week0/cms-sw/cmssw/40383/29763/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Comparison Summary



**Summary**:
 - No significant changes to the logs found
 - Reco comparison results: 25 differences found in the comparisons
 - DQMHistoTests: Total files compared: 49
 - DQMHistoTests: Total histograms compared: 3555748
 - DQMHistoTests: Total failures: 1188
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3554538
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 48 files compared)
 - Checked 211 log files, 162 edm output root files, 49 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_13_0_X_2022-12-22-1100+cb2a0a/54795/triggerResults): no differences found

makortel commented:
Comparison differences are from 11634.7 and thus spurious (#39803)

Dr15Jones commented on FWCore/Framework/interface/ComponentFactory.h:
```diff
@@ -55,31 +55,28 @@ namespace edm {
         std::string modtype = iConfiguration.template getParameter<std::string>("@module_type");
         //cerr << "Factory: module_type = " << modtype << endl;
         typename MakerMap::iterator it = makers_.find(modtype);
+        Maker* maker = nullptr;
```
bare pointer feels `icky`. Maybe use a non owning smart pointer? Also, could this be `const`?

makortel replied:
pThe cache in `Factory` uses `unique_ptr`, whereas the cache in `ComponentFactory` uses `shared_ptr`, so out-of-the box e.g. `weak_ptr` can't be used. On the other hand, I don't see any reason why `Factory` and `ComponentFactory` could not use the same smart pointer.

So I could move both caches to use `shared_ptr` and use `weak_ptr` here, but given that (IIUC) using the `weak_ptr` would be effectively the same as copying the `shared_ptr`, I wonder if it would really be worth it? (even if the module creation is not that performance critical code). 

Making these pointers to pointers-to-`const` seems to work.

Dr15Jones commented on FWCore/Framework/interface/ModuleTypeResolverBase.h:
```diff
@@ -35,9 +38,24 @@ namespace edm {
     virtual std::pair<std::string, int> resolveType(std::string basename, int index) const = 0;
   };
+
+  class ModuleTypeResolverMaker {
```
Shouldn't this be in its own header?

makortel replied:
Done

makortel replied:
Could you check the comment I added in `ModuleTypeResolverMaker.h` makes sense?

Dr15Jones commented on FWCore/Framework/interface/ModuleTypeResolverMakerFactory.h:
```diff
@@ -0,0 +1,15 @@
+namespace edm {
+  class ModuleTypeResolverMaker;
+
+  using ModuleTypeResolverMakerFactory = edmplugin::PluginFactory<ModuleTypeResolverMaker*()>;
```
Could this be `ModuleTypeResolverMaker const*`

makortel replied:
Seems to work, thanks.

Dr15Jones commented on FWCore/Framework/interface/WorkerRegistry.h:
```diff
@@ -37,7 +38,7 @@ namespace edm {
   public:
-    explicit WorkerRegistry(std::shared_ptr<ActivityRegistry> areg);
+    explicit WorkerRegistry(std::shared_ptr<ActivityRegistry> areg, ModuleTypeResolverMaker const* resolverMaker);
     WorkerRegistry(std::shared_ptr<ActivityRegistry> areg, std::shared_ptr<ModuleRegistry> iModReg);
```
Why doesn't this one need `ModuleTypeResolverMaker const*`?

[Answer my own question] Because only the ModuleRegistry needs it and that is being passed in.

Dr15Jones commented on FWCore/Framework/interface/resolveMaker.h:
```diff
@@ -12,28 +13,56 @@ namespace edm::detail {
+  auto resolveMaker(std::string const& moduleType,
+                    ModuleTypeResolverMaker const* resolverMaker,
+                    edm::ParameterSet const& modulePSet,
+                    TCache& makerCache) -> std::tuple<typename TCache::mapped_type::element_type*, bool> {
```
Can the return type every be `{<good value>, false}` or `{nullptr, true}`? If not, then there is no need for the `bool`.

makortel replied:
Good catch. The `{nullptr, false}` is returned only if the insertion of the maker to the cache had failed (so that the caller can thrown an exception). Currently `ComponentFactory` throws an exception if the insertion failed, but `Factory` does not. Maybe it would make most sense to have this `resolveMaker` to throw the exception.

Dr15Jones replied:
Or you can just use 'nullptr' to denote a failure.

makortel replied:
Ended up with returning `nullptr` for failure.

Dr15Jones reviewed (changes requested):
Reviewed all files.

makortel commented on Mixing/Base/src/SecondaryEventProvider.cc:
```diff
@@ -52,7 +52,8 @@ namespace edm {
       : exceptionToActionTable_(new ExceptionToActionTable),
-        workerManager_(std::make_shared<ActivityRegistry>(), *exceptionToActionTable_) {
+        // no type resolver for modules in SecondaryEventProvider for now
+        workerManager_(std::make_shared<ActivityRegistry>(), *exceptionToActionTable_, nullptr) {
```
I left it out because it wasn't clear to me how to percolate the `ModuleTypeResolverMaker const*` to `SecondaryEventProvider` without resorting to global variables (or maybe it should be taken from the "process configuration" the module). I'll create an issue after we've settled with this PR.

makortel replied:
Opened the issue as https://github.com/cms-sw/cmssw/issues/40497

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-40383/33574

- This PR adds an extra 240KB to repository

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  FWCore/Integration/test/BuildFile.xml  modified in PR(s): #40430
  - File  FWCore/ParameterSet/python/Config.py  modified in PR(s): #37996

makortel commented on FWCore/Integration/test/run_TestModuleTypeResolver.sh:
```diff
@@ -0,0 +1,34 @@
+TEST=testModuleTypeResolver_cfg.py
+
+function die { echo Failure $1: status $2 ; exit $2 ; }
+function runSuccess {
```
Maybe? It isn't much though, but felt handy (to reduce clutter and copy paste) when running the same configuration many times with different options. I'd leave for a subsequent PR (if done).

makortel committed:
Make caches to hold pointer-to-const

makortel committed:
Fix wording

makortel commented on FWCore/Integration/test/testModuleTypeResolver_cfg.py:
```diff
@@ -0,0 +1,109 @@
+    iovIsRunNotTime = cms.bool(True)
+)
+
+process.esTestProducerA = cms.ESProducer("ESTestProducerA", valueCpu = cms.int32(10), valueOther = cms.int32(20))
```
Yes, thanks for catching. The test missed it because the `esTestAnalyzerA` was not in the Path. And that revealed a few other problems (in the test setup).

makortel commented on FWCore/ParameterSet/python/Config.py:
```diff
@@ -31,6 +31,7 @@ class edm(object):
     class errors(object):
         #Allowed errors to used within Python
```
Done

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-40383/33581

- This PR adds an extra 240KB to repository

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  FWCore/Integration/test/BuildFile.xml  modified in PR(s): #40430
  - File  FWCore/ParameterSet/python/Config.py  modified in PR(s): #37996

makortel commented:
@cmsbuild, please test

cmsbuild commented:
Pull request #40383 was updated. @makortel, @smuzaffar, @civanch, @Dr15Jones, @mdhildreth can you please check and sign again.

Dr15Jones commented on FWCore/Framework/interface/ModuleTypeResolverMaker.h:
```diff
@@ -0,0 +1,38 @@
+   * the creation of module type resolver objects (derived from
+   * ModuleTypeResolverBase). This pattern allows the use of
+   * information in the module PSet in the module type resolver in a
+   * way that the PSet is parsed only once per module instance.
```
Maybe an example of why one would want to use info in the module PSet?

makortel replied:
How about now?

cmsbuild commented:
Pull request #40383 was updated. @smuzaffar, @civanch, @Dr15Jones, @makortel, @mdhildreth, @cmsbuild can you please check and sign again.

makortel commented:
@cmsbuild, please test

cmsbuild commented:
+1

**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-cb2a0a/29815/summary.html
**COMMIT**: 4b2fb001df07eb95462c6d8007419114ffc434f0
**CMSSW**: CMSSW_13_0_X_2023-01-06-1100/el8_amd64_gcc11
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week0/cms-sw/cmssw/40383/29815/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Comparison Summary



**Summary**:
 - No significant changes to the logs found
 - Reco comparison results: 7 differences found in the comparisons
 - DQMHistoTests: Total files compared: 49
 - DQMHistoTests: Total histograms compared: 3555748
 - DQMHistoTests: Total failures: 157
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3555569
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 48 files compared)
 - Checked 211 log files, 162 edm output root files, 49 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_13_0_X_2023-01-06-1100+cb2a0a/54841/triggerResults): no differences found

makortel commented:
Comparison differences are in 11634.7 (#39803)

makortel commented:
Two small things I had discovered (earlier) in subsequent work

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-40383/33591

- This PR adds an extra 304KB to repository

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  FWCore/Integration/test/BuildFile.xml  modified in PR(s): #40430
  - File  FWCore/ParameterSet/python/Config.py  modified in PR(s): #37996

cmsbuild commented:
Pull request #40383 was updated. @smuzaffar, @civanch, @Dr15Jones, @makortel, @mdhildreth, @cmsbuild can you please check and sign again.

makortel commented:
@cmsbuild, please test

cmsbuild commented:
+1

**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-cb2a0a/29820/summary.html
**COMMIT**: de3e4efa3174b1b54e4ed53169dafc7f81b3b17a
**CMSSW**: CMSSW_13_0_X_2023-01-06-1100/el8_amd64_gcc11
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week0/cms-sw/cmssw/40383/29820/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Comparison Summary



**Summary**:
 - No significant changes to the logs found
 - Reco comparison results: 31 differences found in the comparisons
 - DQMHistoTests: Total files compared: 49
 - DQMHistoTests: Total histograms compared: 3555748
 - DQMHistoTests: Total failures: 117
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3555609
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 48 files compared)
 - Checked 211 log files, 162 edm output root files, 49 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_13_0_X_2023-01-06-1100+cb2a0a/54845/triggerResults): no differences found

civanch commented:
+1

makortel commented:
Hmh, many phase2 workflows show differences in `JetMET/METValidation` (that can't be caused by this PR)

makortel commented:
@cmsbuild, please test

cmsbuild commented:
-1

**Failed Tests**: RelVals-INPUT 
**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-cb2a0a/29846/summary.html
**COMMIT**: de3e4efa3174b1b54e4ed53169dafc7f81b3b17a
**CMSSW**: CMSSW_13_0_X_2023-01-09-1100/el8_amd64_gcc11
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week1/cms-sw/cmssw/40383/29846/install.sh` to create a dev area with all the needed externals and cmssw changes.


## RelVals-INPUT

The relvals timed out after 4 hours.


## Comparison Summary



**Summary**:
 - No significant changes to the logs found
 - Reco comparison results: 33 differences found in the comparisons
 - DQMHistoTests: Total files compared: 49
 - DQMHistoTests: Total histograms compared: 3555538
 - DQMHistoTests: Total failures: 1217
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3554299
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 48 files compared)
 - Checked 211 log files, 162 edm output root files, 49 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_13_0_X_2023-01-09-1100+cb2a0a/54870/triggerResults): no differences found

makortel commented:
This time the comparison differences were concentrated to 11634.7 (#39803). The RelVals-INPUT tests succeeded [earlier](https://github.com/cms-sw/cmssw/pull/40383#issuecomment-1374386701),

makortel commented:
+1

cmsbuild commented:
This pull request is fully signed and it will be integrated in one of the next master IBs (but tests are reportedly failing). This pull request will now be reviewed by the release team before it's merged. @perrotta, @dpiparo, @rappoccio (and backports should be raised in the release meeting by the corresponding L2)

perrotta commented:
+1

perrotta commented:
merge
